pkg/operator/upgradebackupcontroller/upgradebackupcontroller.go

```diff
@@ -0,0 +1,258 @@
+type UpgradeBackupController struct {
+	operatorClient       v1helpers.OperatorClient
+	kubeClient           kubernetes.Interface
+	podLister            corev1listers.PodLister
+	nodeLister           corev1listers.NodeLister
+	networkLister        configv1listers.NetworkLister
+	clusterVersionLister configv1listers.ClusterVersionLister
+	targetImagePullSpec  string
+	operatorVersion      string
+}
+
+func NewUpgradeBackupController(
+	operatorClient v1helpers.OperatorClient,
+	kubeClient kubernetes.Interface,
+	kubeInformers v1helpers.KubeInformersForNamespaces,
+	clusterVersionInformer configv1informers.ClusterVersionInformer,
+	networkInformer configv1informers.NetworkInformer,
+	eventRecorder events.Recorder,
+	targetImagePullSpec string,
+	operatorVersion string,
+) factory.Controller {
+	c := &UpgradeBackupController{
+		operatorClient:       operatorClient,
+		operatorVersion:      operatorVersion,
+		kubeClient:           kubeClient,
+		podLister:            kubeInformers.InformersFor(operatorclient.TargetNamespace).Core().V1().Pods().Lister(),
+		nodeLister:           kubeInformers.InformersFor("").Core().V1().Nodes().Lister(),
+		networkLister:        networkInformer.Lister(),
+		clusterVersionLister: clusterVersionInformer.Lister(),
+		targetImagePullSpec:  targetImagePullSpec,
+	}
+	return factory.New().ResyncEvery(10*time.Minute).WithInformers(
+		kubeInformers.InformersFor(operatorclient.TargetNamespace).Core().V1().Pods().Informer(),
+		kubeInformers.InformersFor("").Core().V1().Nodes().Informer(),
+		operatorClient.Informer(),
+		clusterVersionInformer.Informer(),
+	).WithSync(c.sync).ToController("ClusterBackupController", eventRecorder.WithComponentSuffix("cluster-backup-controller"))
+}
+
+func (c *UpgradeBackupController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	clusterVersion, err := c.clusterVersionLister.Get("cluster")
+	if err != nil {
+		return err
+	}
+
+	// Check cluster version status for backup condition
+	if !isRequireRecentBackup(clusterVersion) {
+		// Work is done
+		return nil
+	}
+
+	// Find the backup condition
+	_, status, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	backupVersion := c.operatorVersion
+	backupCondition := v1helpers.FindOperatorCondition(status.Conditions, backupConditionType)
+	newConditionRequired := backupCondition == nil || backupCondition.Reason != backupVersion
+	if newConditionRequired {
+		_, _, err := v1helpers.UpdateStatus(c.operatorClient, v1helpers.UpdateConditionFn(operatorv1.OperatorCondition{
+			Type:   "RecentBackup",
+			Status: operatorv1.ConditionUnknown,
+			Reason: backupVersion,
+		}))
+		if err != nil {
+			return err
+		}
+	} else if backupCondition.Status != operatorv1.ConditionUnknown {
+		// Backup was taken or failed, but is not in progress.
+		return nil
+	}
+
+	backupStatus, err := ensureRecentBackup(ctx, c.kubeClient, c.podLister, c.nodeLister, backupVersion, c.targetImagePullSpec, syncCtx.Recorder())
```

Comment:
might want to return a condition pointer and `error` here, instead of just the status, so we can create nice reason/message tuples for the various cases.